### PR TITLE
KNX start/stop dimming #635

### PR DESCRIPTION
**Pull request by aruder77**
A solution for issue #635 to support dimming with start/stop telegrams (DPT 3.007).

### CODE REVIEW

**Comment by msfunty** [comment]
I really appreciate that functionality as several KNX interfacing are sending START/STOP DIMMING only instead of the dimmed value.

Within the coding the the period for resending is hard-coded (constant SLEEP_PERIOD_MS). IMHO I would be better to use sendFrequency from Sitemap's definition. Furthermore is there a reason to check max looping?

**Comment by aruder77** [comment]
I agree that the value should not be hard-coded but configurable. The sendFrequency parameter is not usable for the binding though, since it is solely a UI button behaviour thing (at least from my understanding). I'll try to add a config parameter in the binding configuration...
The max looping is only to keep the binding from looping forever in case of an error.

**Comment by cniweb** [comment]
@tobyS Work this PR?

**Comment by teichsta** [comment]
hope to find the time to review until weekend …

**Comment by polarathene** [comment]
@teichsta Any chance to review this PR and merge it? Would satisfy [this 2013 issue](https://github.com/openhab/openhab/issues/635) for it.

File: bundles/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/bus/KNXBinding.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2016, openHAB.org and others.
+ * Copyright (c) 2010-2015, openHAB.org and others.
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/bus/KNXBinding.java]: Should revert this to 2016.

File: bundles/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/bus/KNXBinding.java [hunk 2]
```diff
@@ -470,4 +503,68 @@ private String toDPTValue(Type type, String dpt) {
...
+
+    private void stopDimmerThread(String item) {
+        DimmerThread dimmerThread = itemDimmerThreads.remove(item);
+        if (dimmerThread != null) {
+            dimmerThread.stopRunning();
+        }
+    }
+
+    private void startDimmerThread(GroupAddress destination, String item, Command type) {
+        logger.trace("Starting new dimmer thread for item " + item + ".");
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/bus/KNXBinding.java]: Should use parameterized logging here.

File: bundles/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2016, openHAB.org and others.
+ * Copyright (c) 2010-2015, openHAB.org and others.
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/dpt/KNXCoreTypeMapper.java]: Should revert this to 2016.

**Comment by 9037568** [comment]
LGTM. If noone objects, I'll merge this soon.

**Comment by udo1toni** [comment]
Do I have to configure something to make use of Start-Stop-Dimming? Updated openHAB to 1.8.3 and got addons from here:https://openhab.ci.cloudbees.com/job/openHAB1-Addons/1285/ but dimming through increase/decrease commands doesn't work as expected.

**Comment by aruder77** [comment]
To use it, just add an 'ss' to the group address in the binding configuration, e.g. {knx="3.007:2/0/1ss"}.

**Comment by udo1toni** [comment]
:-( still no luck. Maybe I got the wrong version? As the nightly builds have changed since 1.8.2, where exactly do I have to pull the runtime and the addons from?

**Comment by watou** [comment]
> where exactly do I have to pull the runtime and the addons from

This PR did not change the runtime, just the KNX binding and tests.  Try here:

https://openhab.ci.cloudbees.com/job/openHAB1-Addons/lastSuccessfulBuild/artifact/bundles/binding/org.openhab.binding.knx/target/org.openhab.binding.knx-1.9.0-SNAPSHOT.jar

**Comment by udo1toni** [comment]
So, this pr does not work as expected.
cutout from my .items:
```
Dimmer  Light_FF_Parents_Planets        "Eltern Strahler [%.0f %%]"     (GLights)               { knx="8/5/0+<8/5/6,     8/5/1ss, 8/5/2+<8/5/7" }
Dimmer  Light_FF_Parents_Stars  "Eltern Sterne [%.0f %%]"       (GLights)               { knx="8/5/10+<8/5/16,   8/5/11ss,8/5/12+<8/5/17" }
```
cutout from my .sitemap:
```
Slider item= Light_FF_Parents_Planets   label="Strahler [%.0f %%]"      sendFrequency=400 switchSupport
Slider item= Light_FF_Parents_Stars             label="Sterne [%.0f %%]"        sendFrequency=400 switchSupport
```
If holding down an up/down button, I see INCREASE or DECREASE commands at the log, but the dimmer does not stop dimming at all but reaches maximum or minimum, so there is no additional stop command.

**Comment by aruder77** [comment]
I tested today again and for me it works without any problem. I do not understand your configuration, though. It seems that you are trying to use the start/stop binding the other way around. This PR is a solution to support knx sensors (e.g. touch switches) which only support start/stop dimming like the MDT BE-GT04x01. These touch sensors only send a start telegram when dimming starts and a stop telegram when it stops. Without this PR it is not possible to dimm an openhab item with this sensor.

This is my item and sitemap configuration. With this configuration, I am able to dimm a hue lightbulb using an MDT glass touch sensor (which only supports start/stop dimming).

```
Dimmer Light_GF_Dining_Table_Dimm "Spot 1" <colorlight> { knx="1/0/24, 1/1/22ss", hue="1;brightness;30" }

Slider item=Light_GF_Dining_Table_Dimm
```

Are you trying to control a knx dimming actor which only supports start/stop dimming? I am not aware that devices like that exist...

**Comment by aruder77** [comment]
I'm sorry, I just re-read the original issue and understood that it was actually meant the other way around. I will try to add that functionality but unfortunately I don't have a respective device to test it...

**Comment by udo1toni** [comment]
Ah, I see... Then no wonder, that it doesn't work for me :)

**Comment by mahadi** [comment]
A bit late to the party, but works also for me with Hue :) Thank you! I didn't know about the "ss".

**Pull request merged**